src/workflow_provider/default_vm_workflow_provider.cpp

```diff
@@ -0,0 +1,378 @@
+        }
+        else if (tokens.size() == 1)
+        {
+            query.release = tokens[0];
+        }
+        else
+        {
+            needs_update = true;
+            throw InvalidWorkflowException("Unsupported image scheme in Workflow");
+        }
+    }
+
+    if (workflow_instance["limits"]["min-cpu"])
+    {
+        try
+        {
+            auto min_cpus = workflow_instance["limits"]["min-cpu"].as<int>();
+
+            if (vm_desc.num_cores == 0)
+            {
+                vm_desc.num_cores = min_cpus;
+            }
+            else if (vm_desc.num_cores < min_cpus)
+            {
+                throw WorkflowMinimumException("Number of CPUs", std::to_string(min_cpus));
+            }
+        }
+        catch (const YAML::BadConversion&)
+        {
+            needs_update = true;
+            throw InvalidWorkflowException(fmt::format("Minimum CPU value in workflow is invalid"));
+        }
+    }
+
+    if (workflow_instance["limits"]["min-mem"])
+    {
+        auto min_mem_size_str{workflow_instance["limits"]["min-mem"].as<std::string>()};
+
+        try
+        {
+            MemorySize min_mem_size{min_mem_size_str};
+
+            if (vm_desc.mem_size.in_bytes() == 0)
+            {
+                vm_desc.mem_size = min_mem_size;
+            }
+            else if (vm_desc.mem_size < min_mem_size)
+            {
+                throw WorkflowMinimumException("Memory size", min_mem_size_str);
+            }
+        }
+        catch (const InvalidMemorySizeException&)
+        {
+            needs_update = true;
+            throw InvalidWorkflowException(fmt::format("Minimum memory size value in workflow is invalid"));
+        }
+    }
+
+    if (workflow_instance["limits"]["min-disk"])
+    {
+        auto min_disk_space_str{workflow_instance["limits"]["min-disk"].as<std::string>()};
+
+        try
+        {
+            MemorySize min_disk_space{min_disk_space_str};
+
+            if (vm_desc.disk_space.in_bytes() == 0)
+            {
+                vm_desc.disk_space = min_disk_space;
+            }
+            else if (vm_desc.disk_space < min_disk_space)
+            {
+                throw WorkflowMinimumException("Disk space", min_disk_space_str);
+            }
+        }
+        catch (const InvalidMemorySizeException&)
+        {
+            needs_update = true;
+            throw InvalidWorkflowException(fmt::format("Minimum disk space value in workflow is invalid"));
+        }
+    }
+
+    if (workflow_instance["cloud-init"]["vendor-data"])
+    {
+        try
+        {
+            auto cloud_init_config = YAML::Load(workflow_instance["cloud-init"]["vendor-data"].as<std::string>());
+
+            for (const auto& config : cloud_init_config)
+            {
+                if (config.first.IsScalar())
+                {
+                    vm_desc.vendor_data_config[config.first.Scalar()] = config.second;
+                }
+            }
```

Comment:
This may need to be smarter at some point. Users may need to be part of the workflow's vendor config.

Reply:
> Users may need to be part of the workflow's vendor config.

I'm not sure I understand what you mean here.  Could you please elaborate?

Reply:
I just mean we'll need to be smarter about _adding_ a user, rather than replacing the whole `users:` bit, and such. But later :)

Reply:
Oh, right, add to a list, etc.